[RPC Reliability] Enabled retries for RPCs with exponential backoff

osalpekar opened the pull request:
Summary:
This adds functionality for re-trying RPC's that are sent with the function sendWithRetries(). It adds RPC's that will potentially need to be retried to a sorted map that contains the timeout at which to retry the RPC and associated metadata. A separate thread iteratively removes the earliest retry-able RPC from the map, sleeps until the corresponding time point, re-tries the RPC, and adds to the map again with a future timeout.

GitHub Issue: https://github.com/pytorch/pytorch/issues/32124

Per the first 4 milestones, the following will be addressed in future PR's:
* enabling RPC Retries for RRef internal messages

Differential Revision: D19915694

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -244,6 +244,7 @@ void ProcessGroupAgent::start() {
 void ProcessGroupAgent::shutdown() {
   LOG(INFO) << "Shutting down ProcessGroupAgent on rank " << pg_->getRank()
             << ".";
+  RpcAgent::shutdown();
```
Wondering: does it make more sense to put the superclass's shutdown() method at the end?
This looks functionally like a destructor idiom, in which case one normally expects the subclass to destruct itself before deferring to the superclass's shutdown.

jjlilley replied:
(My comment also applies in parallel to ThriftRpcAgent)

osalpekar replied:
Sounds good, I'll make this change

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -244,6 +244,7 @@ void ProcessGroupAgent::start() {
             << ".";
+  RpcAgent::shutdown();
   std::unique_lock<std::mutex> lock{futureMutex_};
   if (!rpcRunning_.exchange(false)) {
```
btw, reading the code, it's easy to confuse rpcRunning_ from rpcAgentRunning_.
Especially because rpcAgentRunning_ is protected and visible in this scope 

Not sure exactly what to say (maybe just making RpcAgent's var private might help?) but this sort of thing could lead to future bugs, since the var names are easy to confuse

osalpekar replied:
I actually plan on consolidating `rpcRunning_` and `rpcAgentRunning_` in a future PR. Here's the issue to track: #32963

jjlilley replied:
Sounds good. Just be careful how you do exchange(false) during shutdown, especially when superclass's shutdown is at the end.

jjlilley reviewed (commented):
LGTM overall, again.
Just checking - is the main difference in the shutdown logic?

osalpekar commented:
> LGTM overall, again.
> Just checking - is the main difference in the shutdown logic?

Yes, the shutdown logic is the only difference!

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -6,16 +6,182 @@ namespace rpc {
+}
+
+RpcAgent::~RpcAgent() {
+  shutdown();
```
On further thought: Virtual method calls aren't such a great idea in base class destructors.
Can you remove this, and either call in the subclass, or (even better) expect the api user to call it?

The problem-
   1 the ~ProcessGroupAgent (or ThriftRpcAgent) destructor runs.
   2 the ProcessGroupAgent part of the class is effectively destroyed
   3  then ~RpcAgent destructor starts.
   4. It then calls back up into the ProcessGroupAgent::shutdown() method.
    5. This ends up maybe being ok for PGA because the super class implements idempotency. 
        But calling back up into a destroyed subclass's virtual functions tends to end badly.

osalpekar replied:
I'll add a guard to make `ThriftRpcAgent::shutdown()` idempotent. `test_duplicate_name` fails if we don't call `shutdown()` from the RpcAgent destructor. I did some logging to dig into this - the ProcessGroupAgent constructor throws an exception in `ProcessGroupAgent::collectNames()` since two agents have the same workerName. The python test asserts the exception is raised, then it calls the RpcAgent destructor and looks like it never calls the PGA destructor.

jjlilley replied:
- So, this is interesting - I agree that if you write a test program with the current version of clang, it appears to devirtualize the virtual shutdown() call() when called from ~RpcAgent's (maybe swapping the vtables) to avoid calling up to ProcessGroupAgent.
   Does anybody know offhand if this is in the spec, or simply undefined behavior? (though I'll argue below that either way we probably don't want it)

- That said, that means that the shutdown() call from ~RpcAgent() is always effectively broken - it will never call up to ProcessGroupAgent::shutdown(). The non ~RpcAgent shutdown() call is effectively always needed anyway, so there shouldn't be any harm from removing from the ~RpcAgent destructor.

- And honestly, in OSS code, we should be aiming for readability and avoiding error-prone patterns, rather than probing the boundaries of standard and potentially undefined behavior.
   If you Google search "calling virtual function from destructor", there isn't a magic link to the spec up top, but there are a ton of articles telling people to avoid doing this.

My advice here is to minimally avoid calling any virtuals from ~RpcAgent.
I'd also prefer we avoided from ~PGA an ~ThriftRpcAgent as well - though it seems to be much lower risk, since these classes are effectively implicitly final.

rohan-varma replied:
According to https://wiki.sei.cmu.edu/confluence/display/cplusplus/OOP50-CPP.+Do+not+invoke+virtual+functions+from+constructors+or+destructors, we should not be calling virtual functions from constructors/destructors, since it won't result in calling the more derived implementation. I think @jjlilley is right that we do this in `ProcessGroupAgent` and ThriftRpcAgent currently, and ideally we should remove this pattern. 

@osalpekar, is there anyway you could modify `test_invalid_names` not construct/destruct RPC agents at all? I thought it didn't already but as per your logging looks like it does.

osalpekar replied:
`test_duplicate_names` essentially causes PGA to throw an exception in the constructor (due to a call to ProcessGroupAgent::collectNames()). Since this is in the constructor implmentation of the constructor (after the call to the RpcAgent constructor), the RpcAgent is constructed, but the PGA object is not. Thus on destruction, only the RpcAgent destructor is called, and the test fails (times out) if the destructor doesn't call shutdown since the retry thread is not joinable.

osalpekar replied:
I chatted with Pritam offline, and we decided a separate `RpcAgent::cleanup()` function would be the best way to proceed. It satisfies: 1) not calling a virtual function from the RpcAgent destructor 2) still works if PGA or Thrift Agents are not successfully constructed without depending on the child class's specific implementations 3) makes `RpcAgent::shutdown()` a pure virtual function and 4) doesn't require PGA or Thrift `shutdown()` implementations to call `RpcAgent::shutdown`.

jjlilley reviewed (commented):
Can you also make ThriftRpcAgent::shutdown() idempotent, if you feel like you need to protect process_group_agent's shutdown method?

btw, not saying that is strictly necessary, but as a thought exercise - do we actually need to do all this shutdown magic? (e.g. can we simply throw if the user calls > 1 time, or didn't call before the destructor, and just flush out the errors?)

jjlilley commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -132,6 +190,12 @@ class TORCH_API RpcAgent {
+  // the destructor, and ensures that we can gracefully destruct even if the
+  // child class was not successfully constructed without calling a virtual
+  // function.
+  void cleanup();
```
maybe make it private, if there's no need/desire for external entities to call this?

jjlilley reviewed (commented):
This looks nice - and sounds like a good plan!
The only suggestion I'd add at this point is making the new cleanup() call private, unless you need it to be external-facing

osalpekar committed:
[RPC Reliability] Enabled retries for RPCs with exponential backoff (#33365)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33365

This adds functionality for re-trying RPC's that are sent with the function sendWithRetries(). It adds RPC's that will potentially need to be retried to a sorted map that contains the timeout at which to retry the RPC and associated metadata. A separate thread iteratively removes the earliest retry-able RPC from the map, sleeps until the corresponding time point, re-tries the RPC, and adds to the map again with a future timeout.

GitHub Issue: https://github.com/pytorch/pytorch/issues/32124

Per the first 4 milestones, the following will be addressed in future PR's:
* enabling RPC Retries for RRef internal messages

Differential Revision: D19915694

fbshipit-source-id: 04ba88f05ed2ebd69511129045f6d1f305ab98cb

jjlilley reviewed (approved):
LGTM, thanks!

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -653,8 +653,10 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
           Message({}, {}, MessageType::EXCEPTION), ss.str());
-      timedOutFuture.future_->setError(std::string(
-          exceptionMsg.payload().begin(), exceptionMsg.payload().end()));
+      if (!timedOutFuture.future_->hasError()) {
```
@osalpekar Why we need this? 
We shouldn't expect these future pulled out from the map has been used by other threads, right?

osalpekar replied:
I added this since our method of testing involved sending failed RPCs (by calling send and immediately marking the returned future with an error). Checking whether the future already has an error here avoided double-marking the future with an error (which I believe is illegal). There's a diff out to change the failing-sends implementation (currently in ProcessGroupAgentTest.cpp) by just returning a future marked with an error (and never actually calling `ProcessGroupAgent::send`), and with that change, this line could probably be safely removed.

jjlilley replied:
@xush6528 Agree that this is a race condition, since there's a time gap between evaluating and setting.

I ran into this in an unrelated setting (my rpc server async change), and I think it may now be unnecessary for the tests with [pending] D20317773 fix.
